File: frontend/src/lib/apiClient.js
import axios from 'axios';

// Every request/response in this app follows the backend's consistent
// envelope: { success, message, data, meta? } on 2xx, or
// { success:false, message, errors? } on error (see backend
// src/utils/apiResponse.js and src/middleware/error.middleware.js).

export const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:5000/api/v1';

const ACCESS_TOKEN_KEY = 'epharmacy.accessToken';
const REFRESH_TOKEN_KEY = 'epharmacy.refreshToken';

export function getAccessToken() {
  return localStorage.getItem(ACCESS_TOKEN_KEY);
}

export function getRefreshToken() {
  return localStorage.getItem(REFRESH_TOKEN_KEY);
}

export function setTokens({ accessToken, refreshToken } = {}) {
  if (accessToken) localStorage.setItem(ACCESS_TOKEN_KEY, accessToken);
  if (refreshToken) localStorage.setItem(REFRESH_TOKEN_KEY, refreshToken);
}

export function clearTokens() {
  localStorage.removeItem(ACCESS_TOKEN_KEY);
  localStorage.removeItem(REFRESH_TOKEN_KEY);
}

export const apiClient = axios.create({
  baseURL: API_BASE_URL,
  timeout: 20000,
});

apiClient.interceptors.request.use((config) => {
  const token = getAccessToken();
  if (token) {
    config.headers = config.headers || {};
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

// Listeners the AuthContext can subscribe to so it learns immediately when a
// silent refresh fails and the user needs to be signed out.
const sessionExpiredListeners = new Set();
export function onSessionExpired(listener) {
  sessionExpiredListeners.add(listener);
  return () => sessionExpiredListeners.delete(listener);
}

let refreshPromise = null;

async function refreshAccessToken() {
  const refreshToken = getRefreshToken();
  if (!refreshToken) throw new Error('No refresh token available');

  // Use a bare axios call (not apiClient) to avoid recursively triggering
  // the response interceptor below.
  const { data } = await axios.post(`${API_BASE_URL}/auth/refresh-token`, { refreshToken });
  setTokens(data.data);
  return data.data;
}

apiClient.interceptors.response.use(
  (response) => response,
  async (error) => {
    const { response, config } = error;

    if (response && response.status === 401 && !config._retried && !config.url?.includes('/auth/')) {
      config._retried = true;
      try {
        if (!refreshPromise) refreshPromise = refreshAccessToken().finally(() => { refreshPromise = null; });
        const tokens = await refreshPromise;
        config.headers = config.headers || {};
        config.headers.Authorization = `Bearer ${tokens.accessToken}`;
        return apiClient(config);
      } catch (refreshError) {
        clearTokens();
        sessionExpiredListeners.forEach((listener) => listener());
        return Promise.reject(normalizeError(error));
      }
    }

    return Promise.reject(normalizeError(error));
  }
);

export function normalizeError(error) {
  if (error.response) {
    const body = error.response.data || {};
    const err = new Error(body.message || 'Something went wrong. Please try again.');
    err.status = error.response.status;
    err.errors = body.errors;
    return err;
  }
  if (error.request) {
    const err = new Error('Could not reach the server. Please check your connection and try again.');
    err.status = 0;
    return err;
  }
  return error;
}

// Local-storage signed file URLs come back relative to the backend
// (/api/v1/files/signed?...) since that's all the storage provider knows
// about. A cloud storage provider (S3/Supabase) would return an
// already-absolute https:// URL. Either way, this makes it safe to drop
// straight into an <img src>/<a href> regardless of which one it is.
export function toAbsoluteFileUrl(url) {
  if (!url) return url;
  if (/^https?:\/\//i.test(url)) return url;
  return new URL(url, API_BASE_URL).href;
}

// Small helper so call sites can just do `const data = await unwrap(apiClient.get(...))`
export async function unwrap(promise) {
  const res = await promise;
  return res.data.data;
}

export async function unwrapFull(promise) {
  const res = await promise;
  return { data: res.data.data, meta: res.data.meta, message: res.data.message };
}

File: app/lib/network/api_client.dart
import 'dart:async';

import 'package:dio/dio.dart';

import '../core/app_constants.dart';
import 'api_exception.dart';
import 'supabase_auth_client.dart';
import 'token_storage.dart';

typedef SessionExpiredCallback = void Function();

/// The backend's local-storage provider returns file URLs relative to the
/// API origin (e.g. `/api/v1/files/signed?...`) rather than absolute ones —
/// mirrors `toAbsoluteFileUrl` in the web app's lib/apiClient.js.
String toAbsoluteFileUrl(String url) {
  if (url.startsWith('http://') || url.startsWith('https://')) return url;
  final base = Uri.parse(AppConstants.apiBaseUrl);
  final origin = '${base.scheme}://${base.authority}';
  return url.startsWith('/') ? '$origin$url' : '$origin/$url';
}

/// Single Dio instance shared by every service. Attaches the access token
/// to every request, and transparently refreshes + retries once on a 401
/// (mirrors the web app's axios interceptor in lib/apiClient.js), so screens
/// never have to think about token expiry themselves.
class ApiClient {
  ApiClient._internal() {
    _dio = Dio(
      BaseOptions(
        baseUrl: AppConstants.apiBaseUrl,
        connectTimeout: const Duration(seconds: 20),
        receiveTimeout: const Duration(seconds: 20),
        headers: const {'Accept': 'application/json'},
      ),
    );

    _plainDio = Dio(BaseOptions(baseUrl: AppConstants.apiBaseUrl));

    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) async {
          final token = await TokenStorage.instance.getAccessToken();
          if (token != null && token.isNotEmpty) {
            options.headers['Authorization'] = 'Bearer $token';
          }
          handler.next(options);
        },
        onError: (DioException error, handler) async {
          final response = error.response;
          final requestOptions = error.requestOptions;
          final isAuthRoute = requestOptions.path.contains('/auth/');
          final alreadyRetried = requestOptions.extra['retried'] == true;

          if (response?.statusCode == 401 && !isAuthRoute && !alreadyRetried) {
            try {
              final tokens = await _refreshTokens();
              requestOptions.extra['retried'] = true;
              requestOptions.headers['Authorization'] = 'Bearer ${tokens['accessToken']}';
              final cloned = await _dio.fetch(requestOptions);
              return handler.resolve(cloned);
            } catch (_) {
              await TokenStorage.instance.clear();
              _notifySessionExpired();
              return handler.next(error);
            }
          }
          handler.next(error);
        },
      ),
    );
  }

  static final ApiClient instance = ApiClient._internal();

  late final Dio _dio;
  late final Dio _plainDio;

  Dio get dio => _dio;

  Future<Map<String, dynamic>>? _refreshFuture;
  final List<SessionExpiredCallback> _listeners = [];

  void addSessionExpiredListener(SessionExpiredCallback cb) => _listeners.add(cb);

  void removeSessionExpiredListener(SessionExpiredCallback cb) => _listeners.remove(cb);

  void _notifySessionExpired() {
    for (final cb in List<SessionExpiredCallback>.from(_listeners)) {
      cb();
    }
  }

  Future<Map<String, dynamic>> _refreshTokens() {
    _refreshFuture ??= _doRefresh();
    return _refreshFuture!.whenComplete(() => _refreshFuture = null);
  }

  Future<Map<String, dynamic>> _doRefresh() async {
    final refreshToken = await TokenStorage.instance.getRefreshToken();
    if (refreshToken == null || refreshToken.isEmpty) {
      throw ApiException('No refresh token available');
    }
    try {
      final response = await _plainDio.post('/auth/refresh-token', data: {'refreshToken': refreshToken});
      final body = response.data as Map<String, dynamic>;
      final data = body['data'] as Map<String, dynamic>;
      await TokenStorage.instance.saveTokens(
        accessToken: data['accessToken'] as String,
        refreshToken: data['refreshToken'] as String?,
      );
      return data;
    } on DioException {
      // This wasn't a token this backend issued (e.g. a Google sign-in's
      // Supabase refresh token) - ask Supabase to refresh it instead.
      final session = await SupabaseAuthClient.instance.refresh(refreshToken);
      final accessToken = session['access_token'] as String;
      final newRefreshToken = session['refresh_token'] as String?;
      await TokenStorage.instance.saveTokens(accessToken: accessToken, refreshToken: newRefreshToken);
      return {'accessToken': accessToken, 'refreshToken': newRefreshToken};
    }
  }

  /// Runs [request], unwraps the backend's `{ success, message, data }`
  /// envelope, and hands the raw `data` payload to [mapper]. Any failure
  /// (network or server) is normalized into an [ApiException].
  Future<T> unwrap<T>(
    Future<Response<dynamic>> Function(Dio dio) request,
    T Function(dynamic data) mapper,
  ) async {
    try {
      final response = await request(_dio);
      final body = response.data;
      if (body is Map<String, dynamic>) {
        return mapper(body['data']);
      }
      return mapper(body);
    } on DioException catch (e) {
      throw _mapError(e);
    }
  }

  /// Same as [unwrap] but also hands the response's `meta` block (used for
  /// pagination: page/limit/total/totalPages) to [mapper].
  Future<T> unwrapWithMeta<T>(
    Future<Response<dynamic>> Function(Dio dio) request,
    T Function(dynamic data, Map<String, dynamic>? meta) mapper,
  ) async {
    try {
      final response = await request(_dio);
      final body = response.data;
      if (body is Map<String, dynamic>) {
        final meta = body['meta'];
        return mapper(body['data'], meta is Map<String, dynamic> ? meta : null);
      }
      return mapper(body, null);
    } on DioException catch (e) {
      throw _mapError(e);
    }
  }

  ApiException _mapError(DioException e) {
    final response = e.response;
    if (response != null) {
      final body = response.data;
      String message = 'Something went wrong. Please try again.';
      Map<String, dynamic>? errors;
      if (body is Map<String, dynamic>) {
        final serverMessage = body['message'];
        if (serverMessage is String && serverMessage.isNotEmpty) message = serverMessage;
        if (body['errors'] is Map<String, dynamic>) {
          errors = body['errors'] as Map<String, dynamic>;
        }
      }
      return ApiException(message, statusCode: response.statusCode, errors: errors);
    }
    if (e.type == DioExceptionType.connectionTimeout ||
        e.type == DioExceptionType.receiveTimeout ||
        e.type == DioExceptionType.sendTimeout ||
        e.type == DioExceptionType.connectionError) {
      return ApiException('Could not reach the server. Please check your connection and try again.');
    }
    return ApiException(e.message ?? 'Something went wrong. Please try again.');
  }
}
